src/stack-unix/tcpip_stack_socket.ml

```diff
@@ -59,7 +61,7 @@ module V4 = struct
                  Lwt.return_unit) >>= fun () ->
             loop ()
           in
-          loop ())
+          Lwt.pick [ t.switched_off; loop () ] >>= fun `Stopped -> Lwt_unix.close fd)
```

Comment:
Note that `pick` will try to cancel `loop`, which is unreliable. If it's at `recvfrom` then it will probably work, but if it's in `callback` the cancellation might just be silently dropped. You might want to check whether `t.switched_off` is resolved each time around the loop, just before doing the `recvfrom`, to prevent that.

Reply:
Something like:
```ocaml
  let listen_udpv4 t ~port callback =
    if port < 0 || port > 65535 then
      raise (Invalid_argument (err_invalid_port port))
    else
      (* FIXME: we should not ignore the result *)
      Lwt.async (fun () ->
          UDPV4.get_udpv4_listening_fd t.udpv4 port >>= fun fd ->
          let buf = Cstruct.create 4096 in
          let rec loop () =
            (* TODO cancellation *)
            Lwt.catch (fun () ->
                Lwt.pick
                  [ (t.switched_off >>= fun () -> Lwt.return `Stopped)
                  ; (Lwt_cstruct.recvfrom fd buf [] >>= fun (len, sa) ->
                     Lwt.return (`Recv (len, sa))) ] >>= function
                | `Stopped -> Lwt.return `Stopped
                | `Recv (len, sa) ->
                  let buf = Cstruct.sub buf 0 len in
                  (match sa with
                   | Lwt_unix.ADDR_INET (addr, src_port) ->
                     let src = Ipaddr_unix.V4.of_inet_addr_exn addr in
                     let dst = Ipaddr.V4.any in (* TODO *)
                     callback ~src ~dst ~src_port buf >>= fun () ->
                     Lwt.return `Continue
                   | _ -> Lwt.return `Continue))
              (fun exn ->
                 Log.warn (fun m -> m "exception %s in recvfrom" (Printexc.to_string exn)) ;
                 Lwt.return `Continue) >>= function
            | `Continue -> loop () | `Stopped -> Lwt_unix.close fd
          in
          loop ())
```

is better?

Reply:
I was thinking something more like:
```ocaml
if not (Lwt.is_sleepting t.switched_off) then raise Lwt.Canceled;
Lwt.catch (fun () -> ...
```

Reply:
Oh ok :+1:, I will apply such way on all `listen_*` functions.